Code/GraphMol/FileParsers/testMolSupplier.cpp

```diff
@@ -2627,3 +2629,67 @@ TEST_CASE("testGitHub3517") {
   REQUIRE(l > 0);
   REQUIRE(!sdsup.atEnd());
 }
+
+TEST_CASE(
+    "GitHub Issue #9014: SDMolSupplier enters an infinite loop if number of SGroups is incorrect") {
+  constexpr const char *molblock = R"CTAB(
+                    2D
+
+  0  0  0  0  0  0            999 V3000
+M  V30 BEGIN CTAB
+M  V30 COUNTS 8 7 0 0 0
+M  V30 BEGIN ATOM
+M  V30 1 C 1.299038 0.750000 0.000000 0
+M  V30 2 C 1.299038 2.250000 0.000000 0
+M  V30 3 O 2.598076 3.000000 0.000000 0
+M  V30 4 C 2.598076 -0.000000 0.000000 0
+M  V30 5 C 0.000000 0.000000 0.000000 0
+M  V30 6 H 3.897114 0.750000 0.000000 0
+M  V30 7 H 2.598076 4.500000 0.000000 0
+M  V30 8 H 2.598076 -1.500000 0.000000 0
+M  V30 END ATOM
+M  V30 BEGIN BOND
+M  V30 1 1 1 2
+M  V30 2 1 2 3
+M  V30 3 1 1 4
+M  V30 4 1 1 5
+M  V30 5 1 4 6
+M  V30 6 1 3 7
+M  V30 7 1 4 8
+M  V30 END BOND
+M  V30 BEGIN SGROUP
+M  V30 1 SRU 0 ATOMS=(8 1 2 3 4 5 6 7 8)
+M  V30 END SGROUP
+M  V30 END CTAB
+M  END
+$$$$)CTAB";
+
+  v2::FileParsers::SDMolSupplier sdsup;
+  v2::FileParsers::MolFileParserParams p{.strictParsing = false};
+
+  sdsup.setData(molblock, p);
+
+  std::unique_ptr<ROMol> mol;
+  auto parser_thread = std::async(std::launch::async, [&sdsup, &mol]() {
```

Comment:
It's nice to have it, but in general I don't think it's necessary to do this effort to catch things that run forever. If the "loops forever" bug is still there, it will be obvious when running the tests by hand and will result in the CI builds timing out.

If it does seem like something important to do, I guess we could pull this "let a function run for a while before killing it and failing" logic out into a standalone function we could use in multiple places at some point.